rfc/rfc-58/rfc-58.md

```diff
@@ -0,0 +1,100 @@
+ protected String instantTime;
+ protected String fileId;
+ /**
+ * the index of key in parquet rowGroup num.
+ */
+ protected Integer rowGroupNum;
+ }</pre>
+
+b. Number of rowgroup of a Parquet starts from 0 which continously increases util BlockSize reaches `hoodie.parquet.block.size`.  Since every record in parquet belongs to a rowgroup, we can simply use parquet API to locate rowgroup num of new record which needs to be written into corresponding parquet file, and then record rowgroup num into hoodieRecordLocation of each hoodieRecord.  HoodieRecordLocations will be collected into WriteStatus which will be updated to the index on batch.
+
+c. At phase of tagging index, rowgroup num will be queried out, so that they can be used to accelerate updating files.
+
+Concrete flow of upserting is as below:
+
+![4.jpg](4.jpg)
+
+### steps of writing a parquet file on cow
+
+1.  (upserting) data preparing
+
+    At phase of tag indexing, find out `HoodieRecord.currentLocation.rowGroupNum` of updating records, if rowgroup num is empty, record does implicitly not exists, which means current operation is a INSERT, otherwise DELETE or UPDATE. At next, rowgroup nums are used to make grouping by of the updating records so as to collect all rowgroups which should be updated.
+
+![1.jpg](1.jpg)
+
+2.  rowgroup updating
+
+    The process of updating rowgroup is divided into 5 steps.
+
+    1.  Deserializing and decompressing the columns which need to be combined and assembled into a List<Pari<rowKey,Pari<offset,record>> structure,   where offset represents record's row number in rowgroup(every rowgroup's row number starts with zero).
+
+    2.  Using HoodieRecordPayload#getInsertValue to deserialize the upserting data, then invoking HoodieRecordPayload#getInsertValue to combine the updating rows. 
+
+    3.  Converting combined data into column structure, just like `[{"name":"zs","age":10},{"name":"ls","age":20}] ==> {"name":["sz","ls"],"age":[10,20]}`
```

Comment:
I would assume w/ impl, we will decide whether to take this path depending on the payload impl used. we don't want to incur additional overhead for the ones which may not be effective (for eg OverwriteWithLatestAvroPayload, DefaultHoodieRecordPayload)